File: ps_1_to_100/p8.cpp
// In a 1000-digit number, 
// find the adjacent 13 digits that have the maximum product
// print the product
#include <iostream>
#include <cmath>
#include <vector>
#include <fstream>
using namespace std; 

int get_max_product(vector<long long> digits_of_num){
    
    // vector<int> product_digits(digits_of_num.size()-12);
    long long max = 1;
    int num_digits = 13; 
    cout << digits_of_num.size() << endl;
    for (int i=0; i <= digits_of_num.size()-num_digits; i++) {
        long long prod = 1; 
        for (int j=i; j < i+num_digits; j++){
            prod *= digits_of_num[j];
        // cout << j << endl;
        }
        if (prod > max){
            max = prod;
        }
    }
    return max; 
}

int main() {
// you don't really need the number, just the digits; 
// it treats the number in the file as characters
     ifstream inpFile;
     int num_digits = 1000;
     vector<long long> digits(num_digits); 
     char x;  
     inpFile.open("p8_data.txt");
     
     if (!inpFile){
         cerr << "Unable to open the requested file" << endl;
     }
     int sum = 0;
     while (inpFile >> x){
        digits[sum] = (long long)x - '0';
        cout << digits[sum] << " ";
        sum++;
     }
     inpFile.close();
     cout << "\nThe program ended happily with ending index = " << sum << endl; 
     long long max_product;
     max_product = get_max_product(digits);
     cout << max_product << endl;
     return 1;

}

File: ps_1_to_100/p8.py
# // In a 1000-digit number, 
# // find the adjacent 13 digits that have the maximum product
# // print the product

import numpy as np 
import os 
from copy import deepcopy
import operator

def extract_digits(num):
    digits = []
    num = int(num)
    while (num/10 > 0):
        digits.append(deepcopy(num) % 10)
        num //= 10
    digits.reverse()
    return digits

def get_max_product(arr):
    max_prod = []
    # print(arr.size-13+)
    for i in range(len(arr) - 13):
        # print(arr[i+13])
        max_prod.append(np.prod(arr[i:i+13]))
    return max(max_prod)

with open('p8_data.txt') as rfil:
    lines = rfil.readlines()

nums = []
for line  in lines:
    nums.append(line.rstrip())

digits = []
for n in nums:
    digits.extend(extract_digits(n))

# digits = np.array(digits)
print(get_max_product(digits))